contracts/test/UniswapV2RouterStub.sol

```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity <0.9.0;
+
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import "../SignerBondsUniswapV2.sol";
+
+contract UniswapV2RouterStub is IUniswapV2Router {
```

Comment:
I guess it's a broader discussion for everyone be on the same page and agree on our internal standards whether we "hardhat mock" or use the Stubs. I think we can leave it as-is, but since hardhat offers mocking options it seems that everything here could've been mocked in tests. Maybe we should use Stubs only in cases where mocking for some reasons is not possible?

Reply:
Yep, we could use a Waffle mock in that case but I see two reasons why we shouldn't. First, I'd like to check the values of arguments `swapExactETHForTokens` is invoked with. Those arguments are crucial and we can't check them using Waffle's `to.be.calledOnContract` matcher as `@nomiclabs/hardhat-waffle` plugin doesn't support it at the moment. Secondly, I think having `UniswapV2RouterStub` with some kind of real Uniswap contract imitation gives better readability as we have everything in one place. Understanding how the test is supposed to work, is easier in that case, comparing to the alternative situation where we use Waffle mock and have multiple `mock.<method>.returns` scattered among the codebase.

Reply:
Sure, let's leave it as-is as the Stub provides already a lot of functionality. But just want to mention one thing just for the future reference.

> Those arguments are crucial and we can't check them using Waffle's to.be.calledOnContract

Right, but I think there's a trick for this. Functions that return a value can be mocked with exactly the arguments that you expect. If the arguments won't match, then your expected return value won't be returned. We have the event `emit UniswapV2SwapExecuted(amounts);` after the `swapExactETHForTokens(args)` and `amounts` can be checked there, meaning that args for  `swapExactETHForTokens(args)` should match what you've mocked. When we deal with the functions that won't return anything, then hopefully in the future we can rely on `to.be.calledOnContract` as you mentioned.

```
          mockUniswapRouter.mock.swapExactETHForTokens.wtihArgs(
                amountOutMin,
                path,
                assetPool,
                block.timestamp
            ).returns(your_amount)
```
 If the arguments that you pass won't match the expected arguments, then the `your_amount` won't be returned.